contracts/adapters/TargetWeightWrapExtension.sol

```diff
@@ -0,0 +1,382 @@
+            wrapModule.wrap.selector,
+            setToken,
+            rebalanceInfo.reserveAsset,
+            _targetAsset,
+            _reserveUnits,
+            executionParams[_targetAsset].wrapAdapterName,
+            executionParams[_targetAsset].wrapData
+        );
+        invokeManager(address(wrapModule), data);
+
+        (uint256 targetAssetWeight, uint256 reserveWeight) = getTargetAssetAndReserveWeight(_targetAsset);
+        require(targetAssetWeight < executionParams[_targetAsset].maxTargetWeight, "Target asset overweight post-wrap");
+        require(reserveWeight > rebalanceInfo.minReserveWeight, "Reserve asset underweight post-wrap");
+    }
+
+    /**
+     * @notice Unwraps target asset units into the reserve asset.
+     * @dev Must be called when the reserve asset is underweight, rebalancing is enabled, and the caller is an allowed rebalancer.
+     * Ensures that after unwrapping, the target asset does not become underweight and the reserve asset does not become overweight.
+     * @param _targetAsset Address of target asset to unwrap from
+     * @param _targetUnits Units of target asset to unwrap
+     */
+    function unwrap(
+        address _targetAsset,
+        uint256 _targetUnits
+    )
+        external
+        nonReentrant
+        onlyAllowedRebalancer
+    {
+        require(isRebalancingActive, "Rebalancing is not active");
+        require(rebalanceInfo.targetAssets.contains(_targetAsset), "Invalid target asset");
+        require(isReserveUnderweight(), "Reserve asset is not underweight");
+
+        bytes memory data = abi.encodeWithSelector(
+            wrapModule.unwrap.selector,
+            setToken,
+            rebalanceInfo.reserveAsset,
+            _targetAsset,
+            _targetUnits,
+            executionParams[_targetAsset].wrapAdapterName,
+            executionParams[_targetAsset].unwrapData
+        );
+        invokeManager(address(wrapModule), data);
+
+        (uint256 targetAssetWeight, uint256 reserveWeight) = getTargetAssetAndReserveWeight(_targetAsset);
+        require(targetAssetWeight > executionParams[_targetAsset].minTargetWeight, "Target asset underweight post-unwrap");
+        require(reserveWeight < rebalanceInfo.maxReserveWeight, "Reserve asset overweight post-unwrap");
+    }
+
+    /* ========== Operator Functions ========== */
+
+    /**
+     * @notice Sets the reserve asset, target assets, and their associated execution parameters.
+     * @dev Only callable by the operator.
+     * @dev The weights are percentages where 100% equals 1e18.
+     * @param _reserveAsset Address of the reserve asset.
+     * @param _minReserveWeight Minimum allowable weight of the reserve asset.
+     * @param _maxReserveWeight Maximum allowable weight of the reserve asset.
+     * @param _targetAssets Array of target asset addresses.
+     * @param _executionParams Array of execution parameters corresponding to each target asset.
+     */
+    function setTargetWeights(
```

Comment:
If we want to remove a component we would set the min and max weight to zero, right?

Reply:
good catch on extrema - yes i think it makes sense here, but the target asset weight checks in `wrap()` and `unwrap()` will need to be updated from `<, >` to `<=, >=`

Reply:
updated checks and added tests for removing both the component and reserve asset here [`4269d1e`](https://github.com/IndexCoop/index-coop-smart-contracts/pull/182/commits/4269d1ee222d2de41f3a5572ab5e0b2fa28aeda7)